### PR TITLE
AG-11440 - column interface

### PR DESCRIPTION
**Pull request by peterjrreynolds**

### CODE REVIEW

File: community-modules/core/src/columns/columnEventDispatcher.ts [hunk 1]
```diff
@@ -54,16 +54,16 @@ export class ColumnEventDispatcher extends BeanStub {
         this.eventService.dispatchEvent(event);
     }
 
-    public groupOpened(impactedGroups: ProvidedColumnGroup[]): void {
+    public groupOpened(impactedGroups: AgProvidedColumnGroup[]): void {
         const event: WithoutGridCommon<ColumnGroupOpenedEvent> = {
             type: Events.EVENT_COLUMN_GROUP_OPENED,
-            columnGroup: ProvidedColumnGroup.length === 1 ? impactedGroups[0] : undefined,
+            columnGroup: impactedGroups.length === 1 ? impactedGroups[0] : undefined,
```
Comment (StephenCooper) [inline community-modules/core/src/columns/columnEventDispatcher.ts:60]: This was a bug?!

File: community-modules/core/src/entities/agColumn.ts [hunk 2]
```diff
@@ -896,23 +845,9 @@ export class Column<TValue = any> extends BeanStub implements IHeaderColumn<TVal
         return this.colDef.enableRowGroup === true;
     }
 
-    /**
-     * @deprecated v31.1 Use `getColDef().menuTabs ?? defaultValues` instead.
-     */
-    public getMenuTabs(defaultValues: ColumnMenuTab[]): ColumnMenuTab[] {
```
Comment (StephenCooper) [inline community-modules/core/src/entities/agColumn.ts:902]: I am not sure that have the go ahead yet to drop v31 deprecations. I know it was on the table but not confirmed. (Except all the get/sets from gridApi)

File: community-modules/core/src/entities/agColumnGroup.ts [hunk 3]
```diff
@@ -78,8 +80,8 @@ export class ColumnGroup extends BeanStub implements IHeaderColumn {
 
     public checkLeft(): void {
         // first get all children to setLeft, as it impacts our decision below
-        this.displayedChildren!.forEach((child: IHeaderColumn) => {
-            if (child instanceof ColumnGroup) {
+        this.displayedChildren!.forEach((child) => {
+            if (child instanceof AgColumnGroup) {
```
Comment (StephenCooper) [inline community-modules/core/src/entities/agColumnGroup.ts:84]: Should this use `isColumnGroup`?

File: community-modules/core/src/entities/agColumnGroup.ts [hunk 4]
```diff
@@ -231,35 +216,35 @@ export class ColumnGroup extends BeanStub implements IHeaderColumn {
...
 
-    private addLeafColumns(leafColumns: Column[]): void {
-        this.children!.forEach((child: IHeaderColumn) => {
-            if (child instanceof Column) {
+    private addLeafColumns(leafColumns: AgColumn[]): void {
+        this.children!.forEach((child) => {
+            if (isColumn(child)) {
                 leafColumns.push(child);
-            } else if (child instanceof ColumnGroup) {
+            } else if (child instanceof AgColumnGroup) {
```
Comment (StephenCooper) [inline community-modules/core/src/entities/agColumnGroup.ts:233]: `isColumnGroup`?

File: community-modules/core/src/entities/agColumnGroup.ts [hunk 5]
```diff
@@ -298,7 +283,7 @@ export class ColumnGroup extends BeanStub implements IHeaderColumn {
         this.children!.forEach((child) => {
             // never add empty groups
             const emptyGroup =
-                child instanceof ColumnGroup && (!child.displayedChildren || !child.displayedChildren.length);
+                child instanceof AgColumnGroup && (!child.displayedChildren || !child.displayedChildren.length);
```
Comment (StephenCooper) [inline community-modules/core/src/entities/agColumnGroup.ts:286]: `isColumnGroup`

File: community-modules/core/src/entities/agProvidedColumnGroup.ts [hunk 6]
```diff
@@ -137,10 +140,10 @@ export class ProvidedColumnGroup extends BeanStub implements IProvidedColumn, IE
             return;
         }
 
-        this.children.forEach((child: IProvidedColumn) => {
-            if (child instanceof Column) {
+        this.children.forEach((child) => {
+            if (isColumn(child)) {
                 leafColumns.push(child);
-            } else if (child instanceof ProvidedColumnGroup) {
+            } else if (child instanceof AgProvidedColumnGroup) {
```
Comment (StephenCooper) [inline community-modules/core/src/entities/agProvidedColumnGroup.ts:146]: `isColumnGroup`

File: community-modules/core/src/columns/columnEventDispatcher.ts (hunk 1 above)
Comment (peterjrreynolds) [reply community-modules/core/src/columns/columnEventDispatcher.ts:60]: Yes

File: community-modules/core/src/entities/agColumn.ts (hunk 2 above)
Comment (peterjrreynolds) [reply community-modules/core/src/entities/agColumn.ts:902]: This was an "internal" undocumented method that should never have been exposed. It's no longer used internally. Therefore I'm including it in the list of methods that are not making it into the interface.

File: community-modules/core/src/interfaces/iAggFuncService.ts [hunk 7]
```diff
@@ -1,10 +1,10 @@
+import type { AgColumn } from '../entities/agColumn';
 import type { IAggFunc } from '../entities/colDef';
-import type { Column } from '../entities/column';
 
 export interface IAggFuncService {
     addAggFuncs(aggFuncs: { [key: string]: IAggFunc }): void;
     clear(): void;
-    getDefaultAggFunc(column: Column): string | null;
-    getFuncNames(column: Column): string[];
+    getDefaultAggFunc(column: AgColumn): string | null;
```
[the remark this answers is not in the document]
Comment (peterjrreynolds) [reply community-modules/core/src/interfaces/iAggFuncService.ts:7]: `IAggFuncService` is an internal API. It's not exposed to users.

File: community-modules/core/src/interfaces/iMenuFactory.ts [hunk 8]
```diff
@@ -1,24 +1,24 @@
-import type { Column } from '../entities/column';
+import type { AgColumn } from '../entities/agColumn';
```
Comment (StephenCooper) [inline community-modules/core/src/interfaces/iMenuFactory.ts:1]: Should it be external?

File: community-modules/core/src/main.ts [hunk 9]
```diff
@@ -107,9 +107,10 @@ export {
 } from './dragAndDrop/virtualListDragFeature';
 
 // entities
-export { Column, ColumnPinnedType } from './entities/column';
-export { ColumnGroup, ColumnGroupShowType } from './entities/columnGroup';
-export { ProvidedColumnGroup } from './entities/providedColumnGroup';
+export { Column, ColumnPinnedType, ColumnGroup, ProvidedColumnGroup, ColumnGroupShowType } from './interfaces/iColumn';
+export { AgColumn, isColumn } from './entities/agColumn';
+export { AgColumnGroup as AgColumnGroup, isColumnGroup } from './entities/agColumnGroup';
```
Comment (StephenCooper) [inline community-modules/core/src/main.ts:112]: Don't need the `as AgColumnGroup`

File: community-modules/core/src/rendering/features/setLeftFeature.ts [hunk 10]
```diff
@@ -1,49 +1,53 @@
...
+        eCell: HTMLElement,
+        beans: BeanCollection,
+        colsSpanning?: AgColumn[]
+    ) {
         super();
         this.columnOrGroup = columnOrGroup;
         this.eCell = eCell;
         this.ariaEl = this.eCell.querySelector('[role=columnheader]') || this.eCell;
-        this.colsSpanning = colsSpanning;
+        this.colsSpanning = colsSpanning as AgColumn[];
```
Comment (StephenCooper) [inline community-modules/core/src/rendering/features/setLeftFeature.ts:33]: is this cast needed?

File: community-modules/core/src/interfaces/iMenuFactory.ts (hunk 8 above)
Comment (peterjrreynolds) [reply community-modules/core/src/interfaces/iMenuFactory.ts:1]: This is an internal API

File: packages/ag-grid-charts-enterprise/src/main-styles.ts [hunk 11]
```diff
@@ -107,11 +110,9 @@ export { CheckboxCellEditor } from '@ag-grid-community/core';
 export { CheckboxCellRenderer } from '@ag-grid-community/core';
 export { CheckboxSelectionComponent } from '@ag-grid-community/core';
 export { ClientSideRowModelSteps } from '@ag-grid-community/core';
-export { Column } from '@ag-grid-community/core';
```
Comment (StephenCooper) [inline packages/ag-grid-charts-enterprise/src/main-styles.ts:110]: Shouldn't this be exported?

**Review by StephenCooper — APPROVED** [review]
Looks good apart from comments made. 

Overall I guess there is no way to get around so much casting which is a shame.

File: packages/ag-grid-charts-enterprise/src/main-styles.ts (hunk 11 above)
Comment (peterjrreynolds) [reply packages/ag-grid-charts-enterprise/src/main-styles.ts:110]: This is auto-generated. I believe it doesn't export interfaces.

**Pull request merged**